### PR TITLE
Refactor session mechanism

### PR DESCRIPTION
**Pull request by kiootic**
ref #1241

### CODE REVIEW

File: pkg/auth/dependency/session/context.go [hunk 1]
```diff
@@ -1,5 +1,38 @@
...
 
+import (
+	"context"
+
+	"github.com/skygeario/skygear-server/pkg/core/auth/authinfo"
+)
+
 type Context interface {
 	Session() *Session
+	User() *authinfo.AuthInfo
```
Comment (louischan-oursky) [inline pkg/auth/dependency/session/context.go]: Are we going to replace AuthContextGetter with this interface?

File: pkg/auth/dependency/session/cookie.go [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+
+	corehttp.UpdateCookie(rw, cookieSession)
+}
+
+func ClearCookie(rw http.ResponseWriter, useInsecureCookie bool) {
+	corehttp.UpdateCookie(rw, &http.Cookie{
+		Name:     CookieName,
+		Path:     "/",
+		HttpOnly: true,
+		Secure:   !useInsecureCookie,
```
Comment (louischan-oursky) [inline pkg/auth/dependency/session/cookie.go]: Though it has no effect, we should better keep `SameSite: http.SameSiteLaxMode` here.

File: pkg/auth/dependency/session/context.go (hunk 1 above)
Comment (kiootic) [reply pkg/auth/dependency/session/context.go]: Yes, I planned to have three interface for auth contexts:
- `session.Context`: contains user & session, used in auth gear only (in auth)
- `auth.AccessKeyContext`: contains API key (i.e. client ID), used in needed places (in core)
- `authn.Context`: contains `authn.Info`, used in everywhere (in core)

File: pkg/auth/handler/session/resolve.go [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+	"github.com/skygeario/skygear-server/pkg/auth"
+	"github.com/skygeario/skygear-server/pkg/auth/dependency/session"
+)
+
+func AttachResolveHandler(
+	router *mux.Router,
+	authDependency auth.DependencyMap,
+) {
+	router.NewRoute().
+		Path("/session/get").
```
Comment (carmenlau) [inline pkg/auth/handler/session/resolve.go]: Should be `/session/resolve`?

File: pkg/auth/dependency/session/provider_impl.go [hunk 4]
```diff
@@ -59,7 +63,6 @@ func (p *ProviderImpl) MakeSession(authnSess *auth.AuthnSession) (*Session, stri
 	// if session has new fields.
 	session := &Session{
 		ID:                      uuid.New(),
-		ClientID:                authnSess.ClientID,
```
[the remark this answers is not in the document]
Comment (kiootic) [reply pkg/auth/dependency/session/provider_impl.go]: This is the IdP session, which is unrelated to specific client. We'll have a separate Access Grant struct to associate access token with specific client.

**Pull request merged**